v0.21: Facets and filters (#1041)
meilisearch/documentation: reference/features/filtering_and_faceted_search.md

# Filtering and faceted search

> You can use MeiliSearch's filters to refine search results and create faceted search interfaces.

Author: I really like your second paragraph, but would prefer to keep the first as it is, if that's ok with you. 

My intention would be to:

1. use the first paragraph to give a brief, general and fairly non-technical overview of what users can do with filters;
2. and then use the second paragraph to hammer down the difference between filtering and faceting.

```suggestion
You can use MeiliSearch's filters to refine search results and create faceted search interfaces.

While filters are more general conditions and comparisons (e.g. `age > 18`), faceted search is a type of filter that focuses on categorizing results (e.g. all documents where `color` is `red` or `blue`).
```

## Configuring filters

Filters use [document fields](/learn/core_concepts/documents.md#fields) to establish filtering criteria.

To use a document field as a filter, you must first add its attribute to the [`filterableAttributes` index setting](/reference/api/filterable_attributes.md).

> **This step is mandatory and cannot be done at search time.** Filters need to be properly processed and prepared by MeiliSearch before they can be used, an operation that takes as much time as indexing all your documents.

Author: ```suggestion
**This step is mandatory and cannot be done at search time.** Filters need to be properly processed and prepared by MeiliSearch before they can be used.

Updating `filterableAttributes` requires recreating the entire index. This may take a significant amount of time depending on your dataset size.
```
```

:::

Filters work with numeric and string values. Empty fields or fields containing an empty array will be ignored. Fields containing object values will throw an error.

### Example

Suppose you have a collection of movies containing the following fields:

```json
[
  {
      "id": "458723",
      "title": "Us",
      "director": "Jordan Peele",
      "genres": [
        "Thriller",
        "Horror",
        "Mystery"
      ],
      "overview": "Husband and wife Gabe and Adelaide Wilson take their […]",
  },
  …
]
```

If you want to filter results based on the `director` and `genre` attributes, you must add them to the [`filterableAttributes` list](/reference/api/filterable_attributes.md).

<CodeSamples id="faceted_search_update_settings_1" />

## Using filters

Once you have configured `filterableAttributes`, you can start using [the `filter` search parameter](/reference/features/search_parameters#filter). Search parameters are added to a query when a user searches for a document.

`filter` expects a **filter expression** containing one or more **conditions**. A filter expression can be written as a string, as an array, or as a mix of both.

> ::: warning
> `filter` is a new search parameter introduced in MeiliSearch v0.21. It combines functionality previously present in v0.20's `filters` and `facetFilters`.
> 
> Both `filters` and `facetFilters` have been removed from v0.21.
> :::

Reviewer: Since we are not in v1 we do not provide support for the previous MeilISearch versions. To avoid additional maintaining I think it should be removed.
```suggestion
```

String expressions combine conditions using three logical connectives—`AND`, `OR` and `NOT`— and parentheses:

- `NOT` only returns documents that do not satisfy a condition : `NOT genres = horror`
- `AND` operates by connecting two conditions and only returns documents that satisfy both of them: `genres = horror AND director = 'Jordan Peele'`
- `OR` connects two conditions and returns results that satisfy at least one of them: `genres = horror OR genres = comedy`

::: tip
String expressions are read left to right. `NOT` takes precedence over `AND` and `AND` takes precedence over `OR`. You can use parentheses to ensure expressions are correctly parsed.

For instance, if you want your results to only include `comedy` and `horror` movies released after March 1995, the parentheses in the following query are mandatory:

`(genre = horror OR genre = comedy) AND release_date > 795484800`

Failing to add these parentheses will cause the same query to be parsed as:

`genre = horror OR (genre = comedy AND release_date > 795484800)`

Translated into English, the above expression will only return comedies released after March 1995 or horror movies regardless of their `release_date`.
:::

> 

Reviewer: You can add that when using `GET /search` only the expression with strings works.
So not arrays and no mixed

#### Creating filter expressions with arrays

> Array expressions establish logical connectives by nesting arrays of strings. Array expressions can have a maximum depth of two—arrays with three or more levels of nesting will throw an error.

Reviewer: ```suggestion
Array expressions establish logical connectives by nesting arrays of strings. Array expressions can have a maximum depth of two—arrays, with three or more levels of nesting will throw an error.
```

Author: This comma would split the subject of the sentence in two (`arrays with three or more levels of nesting` would become a lonely `arrays` followed by `with three or more levels of nesting`), making it a bit confusing.

Just so we're clear: it's hard to see if you're only reviewing the diff, but that dash is not a hyphen, it's an em dash which in this case works like a mix between parentheses and a colon (`:`).

https://deploy-preview-1041--distracted-wiles-3edfc5.netlify.app/reference/features/filtering_and_faceted_search.html#creating-filter-expressions-with-arrays

```
"(genres = comedy OR genres = horror) AND director != 'Jordan Peele'"
```

You can write the same filter using mixing arrays and strings:

```
[["genres = comedy, genres = horror"], "NOT director = 'Jordan Peele'"]
```

### Example

Suppose that you have dataset containing several movies in the following format:

```json
[
    …
    {
        "id": "458723",
        "title": "Us",
        "director": "Jordan Peele",
        "poster": "https://image.tmdb.org/t/p/w1280/ux2dU1jQ2ACIMShzB3yP93Udpzc.jpg",
        "overview": "Husband and wife Gabe and Adelaide Wilson take their…",
        "release_date": 1552521600,
        "genres": [
            "Comedy",
            "Horror",
            "Thriller"
        ],
        "rating": 4
    },
    …
]
```

If you want to enable filtering using `director`, `release_date`, `genres`, and `rating`, you must add these attributes to the [`filterableAttributes` index setting](/reference/api/filterable_attributes.md).

You can then restrict a search so it only returns movies released after 18 March 1995 with the following filter containing a single condition:

```SQL

> release_data > 795484800

Reviewer 2: ```suggestion
release_date > 795484800
```

```

You can use this filter when searching for recent `Avengers` movies:

<CodeSamples id="filtering_guide_1" />

You can also combine multiple conditions. For instance, you can limit your search so it only includes recent movies directed by either `Tim Burton` or `Christopher Nolan`:

```SQL
release_date > 795484800 AND (director = "Tim Burton" OR director = "Christopher Nolan")
```

Here, the parentheses are mandatory: without them, the filter would return movies directed by `Tim Burton` and released after 1995 or any film directed by `Christopher Nolan`, without constraints on its release date. This happens because `AND` takes precedence over `OR`.

You can use this filter when searching for `Batman` movies:

<CodeSamples id="filtering_guide_2" />

Note that filtering on string values is case-insensitive.

If you only want well-rated movies that weren't directed by `Tim Burton`, you can use this filter:

```SQL
rating >= 3 AND (NOT director = "Tim Burton")
```

Querying on `Planet of the Apes`, the above example results in the following command:

<CodeSamples id="filtering_guide_4" />

## Faceted search

> **Faceted search** is a feature provided out-of-the-box by MeiliSearch. Faceting allows classifying search results into categories or **facets**. Facets are essentially a type of filter.

Reviewer: ```suggestion
**Faceted search** is a sub-filter provided by the filter feature. It was briefly discussed in the previous category. As it is a common feature used in search, the following will explain in depth what it is and how to use it.

Faceting allows classifying search results into categories or **facets**. Facets are essentially a type of filter.

example: In my clothes dataset I want all paints that are either bleu or yellow.
```

Author: Coming from my attempt to clarify that facets are a specific usage of filters, based mostly on how filter conditions/criteria are displayed:

```suggestion
MeiliSearch filters can be used to build **faceted search** interfaces, a specific type of interface that allows a user to refine search results based on broad categories or **facets**. For example, a clothing webshop can use faceted search to allow users to easily explore items in a specific color or belonging to a specific brand.
```

Faceted search provides users with a quick way to narrow down search results by selecting categories relevant to what they are looking for. A faceted navigation system is an **intuitive interface to display and navigate through content**. Facets are used in the UI as filters which users can apply to refine the results in real-time.

This is common in ecommerce sites like Amazon: when users perform a search, they are presented not only with a list of results, but also with a list of facets which you can see on the sidebar in the image below:

![Amazon UI](/faceted-search/facets-amazon.png)

### Filters or facets

> Since MeiliSearch v0.21, there is no difference between facets and filters. They function and behave in the exact same way. The question of whether something is a filter or a facet is mostly one pertaining to UX and UI design.

Reviewer: This need to be reworded as facets are a sub-category of filters

Reviewer 2: Rather than a sub-category, I think we should say that facetting is a way to filter from a user experience point of view. The term sub category doesn't seem very clear to me for a user.

Also, do we want to mention the version number?

Author: > Rather than a sub-category, I think we should say that facetting is a way to filter from a user experience point of view. The term sub category doesn't seem very clear to me for a user.

I agree with this and it was probably the biggest challenge in the whole article: as far as I understand it, facets are not a different _type_ of filter, they are a different _use_ of filters.

Author: ```suggestion
### Filters or facets

In MeiliSearch, facets are a specific use-case of filters. The question of whether something is a filter or a facet is mostly one pertaining to UX and UI design.
```

To use a document field as a filter, you must first add its attribute to the [`filterableAttributes` index setting](/reference/api/filterable_attributes.md).

**This step is mandatory and cannot be done at search time.** Filters need to be properly processed and prepared by MeiliSearch before they can be used.

Updating `filterableAttributes` requires recreating the entire index. This may take a significant amount of time depending on your dataset size.

::: warning
By default, `filterableAttributes` is empty. This means that filters do not work without first explicitly adding attributes to the `filterableAttributes` list.
:::

Filters work with numeric and string values. Empty fields or fields containing an empty array will be ignored. Fields containing object values will throw an error.

### Example

Suppose you have a collection of movies containing the following fields:

```json
[
  {
      "id": "458723",
      "title": "Us",
      "director": "Jordan Peele",
      "genres": [
        "Thriller",
        "Horror",
        "Mystery"
      ],
      "overview": "Husband and wife Gabe and Adelaide Wilson take their […]",
  },
  …
]
```

> If you want to filter results based on the `director` and `genre` attributes, you must add them to the [`filterableAttributes` list](/reference/api/filterable_attributes.md):

Reviewer 2: ```suggestion
If you want to filter results based on the `director` and `genres` attributes, you must add them to the [`filterableAttributes` list](/reference/api/filterable_attributes.md):
```

:::

### Filter expressions

You can build filter expressions by grouping basic conditions. Filter expressions can be written as strings, arrays, or a mix of both.

::: warning
The [`GET` route of the search endpoint](/reference/api/search.md#search-in-an-index-with-get-route) only accepts string filter expressions.
:::

#### Creating filter expressions with strings

String expressions combine conditions using three logical connectives—`AND`, `OR` and `NOT`— and parentheses:

- `NOT` only returns documents that do not satisfy a condition : `NOT genres = horror`
- `AND` operates by connecting two conditions and only returns documents that satisfy both of them: `genres = horror AND director = 'Jordan Peele'`
- `OR` connects two conditions and returns results that satisfy at least one of them: `genres = horror OR genres = comedy`

::: tip
String expressions are read left to right. `NOT` takes precedence over `AND` and `AND` takes precedence over `OR`. You can use parentheses to ensure expressions are correctly parsed.

For instance, if you want your results to only include `comedy` and `horror` movies released after March 1995, the parentheses in the following query are mandatory:

> `(genre = horror OR genre = comedy) AND release_date > 795484800`

Reviewer 2: ```suggestion
`(genres = horror OR genres = comedy) AND release_date > 795484800`
```

Failing to add these parentheses will cause the same query to be parsed as:

> `genre = horror OR (genre = comedy AND release_date > 795484800)`

Reviewer 2: ```suggestion
`genres = horror OR (genres = comedy AND release_date > 795484800)`
```
